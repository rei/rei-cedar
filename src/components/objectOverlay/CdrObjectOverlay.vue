<script setup lang="ts">
import { useCssModule, computed } from 'vue';
import mapClasses from '../../utils/mapClasses';

/** Component for positioning content in 9 different positions relative to a container */

defineOptions({ name: 'CdrObjectOverlay' });

interface ObjectOverlayProps {
  /** Position of the content relative to the container */
  position?:
    | 'top-left'
    | 'top-center'
    | 'top-right'
    | 'left'
    | 'center'
    | 'right'
    | 'bottom-left'
    | 'bottom-center'
    | 'bottom-right';
  /** Margin space around the positioned content */
  margin?: string;
  /** Sets the HTML tag for the container element */
  tag?: string;
}

const props = withDefaults(defineProps<ObjectOverlayProps>(), {
  position: 'center',
  margin: '0',
  tag: 'div',
});

const style = useCssModule();

const positionClass = computed(() => `cdr-object-overlay--${props.position}`);
const containerStyle = computed(() => ({
  '--cdr-object-overlay-margin': props.margin,
}));
</script>

<template>
  <component
    :is="tag"
    :class="mapClasses(style, 'cdr-object-overlay', positionClass)"
    :style="containerStyle"
  >
    <div :class="style['cdr-object-overlay__container']">
      <!-- @slot Container content that the overlay will be positioned relative to -->
      <slot name="container" />
    </div>
    <div :class="style['cdr-object-overlay__content']">
      <!-- @slot Content to be positioned -->
      <slot name="content" />
    </div>
  </component>
</template>

<style lang="scss" module src="./styles/CdrObjectOverlay.module.scss" />