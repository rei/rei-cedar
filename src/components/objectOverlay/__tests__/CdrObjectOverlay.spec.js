import { mount } from '../../../../test/vue-jest-style-workaround.js';
import CdrObjectOverlay from '../CdrObjectOverlay.vue';

describe('CdrObjectOverlay.vue', () => {
    describe('default rendering', () => {
      let wrapper;
      let elem;
  
      beforeEach(() => {
        elem = document.createElement('div');
        document.body.appendChild(elem);
        wrapper = mount(CdrObjectOverlay, {
          props: {},
          slots: {
            container: '<div class="container-slot">Container Content</div>',
            content: '<div class="content-slot">Overlay Content</div>',
          },
          attachTo: elem,
        });
      });
  
      afterEach(() => {
        wrapper.unmount();
        elem.remove();
      });
  
      it('renders correctly with default props', () => {
        // Default tag is "div"
        expect(wrapper.element.tagName).toBe('DIV');
        // Check default classes: base class and position "center"
        expect(wrapper.classes()).toContain('cdr-object-overlay');
        expect(wrapper.classes()).toContain('cdr-object-overlay--center');
        // Verify that the inline style includes the default margin (0)
        expect(wrapper.attributes('style')).toMatch(/--cdr-object-overlay-margin:\s*0/);
        // Verify that the "container" and "content" slots are rendered properly
        const containerSlot = wrapper.find('.cdr-object-overlay__container');
        expect(containerSlot.exists()).toBe(true);
        expect(containerSlot.text()).toContain('Container Content');
  
        const contentSlot = wrapper.find('.cdr-object-overlay__content');
        expect(contentSlot.exists()).toBe(true);
        expect(contentSlot.text()).toContain('Overlay Content');
      });
  
      it('matches the snapshot', () => {
        expect(wrapper.element).toMatchSnapshot();
      });
    });
  
    describe('prop handling', () => {
      it('applies the correct class for each custom position', () => {
        const positions = [
          'top-left',
          'top-center',
          'top-right',
          'left',
          'center',
          'right',
          'bottom-left',
          'bottom-center',
          'bottom-right',
        ];
        positions.forEach((position) => {
          const wrapper = mount(CdrObjectOverlay, {
            props: { position },
            slots: {
              container: '<div>Container</div>',
              content: '<div>Content</div>',
            },
          });
          expect(wrapper.classes()).toContain(`cdr-object-overlay--${position}`);
          wrapper.unmount();
        });
      });
  
      it('applies a custom margin through inline style', () => {
        const marginValue = '20px';
        const wrapper = mount(CdrObjectOverlay, {
          props: { margin: marginValue },
          slots: {
            container: '<div>Container</div>',
            content: '<div>Content</div>',
          },
        });
        expect(wrapper.attributes('style')).toMatch(new RegExp(`--cdr-object-overlay-margin:\\s*${marginValue}`));
        wrapper.unmount();
      });
  
      it('uses a custom HTML tag when the "tag" prop is specified', () => {
        const wrapper = mount(CdrObjectOverlay, {
          props: { tag: 'section' },
          slots: {
            container: '<div>Container</div>',
            content: '<div>Content</div>',
          },
        });
        expect(wrapper.element.tagName).toBe('SECTION');
        wrapper.unmount();
      });
    });
  });