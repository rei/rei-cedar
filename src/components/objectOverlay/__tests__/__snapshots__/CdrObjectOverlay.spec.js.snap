// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CdrObjectOverlay.vue > default rendering > matches the snapshot 1`] = `
<div
  class="cdr-object-overlay cdr-object-overlay--center"
  style="--cdr-object-overlay-margin: 0;"
>
  <div
    class="cdr-object-overlay__container"
  >
    <!-- @slot Container content that the overlay will be positioned relative to -->
    
    <div
      class="container-slot"
    >
      Container Content
    </div>
    
  </div>
  <div
    class="cdr-object-overlay__content"
  >
    <!-- @slot Content to be positioned -->
    
    <div
      class="content-slot"
    >
      Overlay Content
    </div>
    
  </div>
</div>
`;
